### PR TITLE
Add System.Text.Json converter for EventGridEvent

### LINKED ISSUE #21639: Add custom converter for EventGridEvent
CloudEvent has a custom System.Text.Json converter but users would need to rely on documentation to know this, and it doesn't support other serializers.

EventGridEvent doesn't have a custom converter so there is not an easy way to serialize to JSON (since the Data property is BinaryData which does not get serialized by JSON serializers).

Proposal:
Add Serialize/SerializeAsync methods to CloudEvent/EventGridEvent that have an optional ObjectSerializer parameter.
Add custom STJ converter to EventGridEvent.

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Fixes https://github.com/azure/azure-sdk-for-net/issues/21639

### CODE REVIEW

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridEventConverter.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+using Azure.Core;
+using Azure.Messaging.EventGrid.Models;
+
+namespace Azure.Messaging.EventGrid
+{
+    /// <summary>
+    /// A custom converter that attributes the <see cref="EventGridEvent"/> type.
+    /// This allows System.Text.Json to serialize and deserialize EventGridEvent by default.
+    /// </summary>
+    internal class EventGridEventConverter : JsonConverter<EventGridEvent>
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridEventConverter.cs:16]: This is tested by existing tests after switching the Parse and Publish methods to use this.

File: sdk/eventgrid/Azure.Messaging.EventGrid/tests/EventGridEventTests.cs [hunk 2]
```diff
@@ -140,7 +140,7 @@ public async Task RespectsPortFromUriSendingEventGridEvents()
 
         private static List<EventGridEvent> DeserializeRequest(Request request)
         {
-            var content = request.Content as Utf8JsonRequestContent;
+            var content = request.Content;
             var stream = new MemoryStream();
```
Comment (jsquire) [inline sdk/eventgrid/Azure.Messaging.EventGrid/tests/EventGridEventTests.cs:144]:
```suggestion
            using var stream = new MemoryStream();
```

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridEventConverter.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        /// <summary>
+        /// Gets or sets the serializer to use for the data portion of the <see cref="EventGridEvent"/>. If not specified,
+        /// JsonObjectSerializer is used.
+        /// </summary>
+        /// <inheritdoc cref="JsonConverter{CloudEvent}.Read(ref Utf8JsonReader, Type, JsonSerializerOptions)"/>
```
Comment (jsquire) [inline sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridEventConverter.cs:22]: Can we do a partial inherit like this?   I've never seen that pattern before.

**Review by jsquire — APPROVED** [review]

File: sdk/eventgrid/Azure.Messaging.EventGrid/tests/EventGridEventTests.cs (hunk 2 above)
Comment (JoshLove-msft) [reply sdk/eventgrid/Azure.Messaging.EventGrid/tests/EventGridEventTests.cs:144]: not necessary as it is a memory stream.

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridEventConverter.cs (hunk 3 above)
Comment (JoshLove-msft) [reply sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridEventConverter.cs:22]: oops

File: sdk/eventgrid/Azure.Messaging.EventGrid/tests/EventGridEventTests.cs (hunk 2 above)
Comment (jsquire) [reply sdk/eventgrid/Azure.Messaging.EventGrid/tests/EventGridEventTests.cs:144]: if disposable it is,  dispose it, you should.

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridEvent.cs [hunk 4]
```diff
@@ -142,16 +144,11 @@ public static EventGridEvent[] ParseMany(BinaryData json)
             if (requestDocument.RootElement.ValueKind == JsonValueKind.Object)
             {
                 egEvents = new EventGridEvent[1];
-                egEvents[0] = (new EventGridEvent(EventGridEventInternal.DeserializeEventGridEventInternal(requestDocument.RootElement)));
+                egEvents[0] = JsonSerializer.Deserialize<EventGridEvent>(json.ToMemory().Span);
```
Comment (pakrym) [inline sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridEvent.cs:147]: What you had was more efficient

**Review by pakrym — APPROVED** [review]

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridEvent.cs (hunk 4 above)
Comment (JoshLove-msft) [reply sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridEvent.cs:147]: Yeah but this validates the converter.

**Pull request merged**